### ISSUE TITLE
[Enhancement]: Support for craft-neo

### ISSUE THREAD

**Issue by stephanedemotte** [markhuot/craftql#86]
Neo for craft 3 will be release soon, a support could be great
[Neo - Craft 3 -> branch dev-3-port](https://github.com/benjamminf/craft-neo/tree/dev-3-port)

**Comment by timkelty** [comment]
Interest in this one too…in the past I've resisted relying on Neo, but in some cases it is just too powerful to ignore 😛

**Comment by timkelty** [comment]
Briefly looked this morning, and I don't think it should be too bad to implement.

Since we can use https://github.com/markhuot/craftql/blob/master/src/Listeners/GetMatrixFieldSchema.php as a starting point, basically we need to be able to do the same thing, except this bit needs to be recursive, based on the Neo Field's `benf\neo\models::BlockStructure`: https://github.com/benjamminf/craft-neo/blob/dev-3-port/src/models/BlockStructure.php

This is what allows Neo to nest child blocks within others.

**Comment by edolyne** [comment]
@timkelty Wanted to bump this one as the Neo plugin has been released for Craft 3.

**Comment by bianpratama** [comment]
@markhuot is there any plan for this feature request?

**Comment by pr1ntr** [comment]
This would be huge for our project

**Comment by markhuot** [comment]
Working on this over here, https://github.com/spicywebau/craft-neo/issues/173

**Comment by lastgiven** [comment]
Is there any Documentation in regards to grabbing data from neo in CraftQL or is this still not in production? 

My Neo Fields are shown as a String so just want to confirm this.

**Comment by lastgiven** [comment]
@markhuot Sorry for the spam. But I need to make a final desition today if I am going to use CraftQl or element API.

**Comment by narration-sd** [comment]
@lastgiven et al, I think if you want Neo, you'll need to contact the folks who've taken it over -- https://github.com/spicywebau/craft-neo.

They'd need to make the connections for CraftQL themselves, as they're responsible for it.  You can read about this on the front page of the CraftQL site, here: https://github.com/markhuot/craftql#third-party-field-support.

Neo's 'freedom to go wild' may seem attractive, but it also probably means connections like CraftQL, and also just seeing it all work, can be a bit challenging. You might want to consider more straightforward data design than Neo implies.

**Comment by lastgiven** [comment]
@narration-sd Yeah, I get what you are saying. But @markhuot contributed to the NEO plugin a while ago to get it working and that pull request of his got removed for some reason. 

See here: https://github.com/spicywebau/craft-neo/issues/173

So that is why I asked here if there is some sort of feedback in regards to this seeing that the discussion between the tow plugin builders already started

**Comment by narration-sd** [comment]
Hmm, yes, an amount of activity over there, and some hacks that apparently worked in a simple fashion, or for a while.

I think this gets at the root of it, though. Neo was an attempt to wire around Craft's viewpoint on sensible design, while Bend long resisted bringing its apparent function inside. The guy who build it faded as Craft moved on without it tieing their hands, walked away. 

Recently, Matrix itself has had a string of issues as Craft progresses its abilities, which suggest the real complexity dangers, which likely would get expanded upon by a Neo. Anything like this would also get more uncomfortable as it became necessarily wrapped up in Project Config, another issue-visible item of the moment. 

So again, basic considerations seem pretty present. I wasn't too pleased with what I've seen people do in ElementApi, to use a 'free-form' data design Neo seemed to make attractive. 

Besides complex, it wasn't at all to EApi's inner design either, making another fragility into futures.

For  this and other reasons, I've been feeling at least here that going with CraftQL, and the more straightforward data access it supports, is going to be the better structuring basis to work from  -- designing your app to go its ways leading to less hair-pulling, and more solid, supportable deliveries.

**Comment by narration-sd** [comment]
As a p.s., I should say that my positive views on CraftQL are probably influenced also by long  work on a plugin I'm about to put into public, which moved on from ElementApi quite a while ago.

**Comment by lastgiven** [comment]
I agree with 100%. 

Unfortunately, my hands are tied at the moment when it comes to using NEO. 

We are busy upgrading an existing project over to Craft 3 where we used NEO as one of the major points in the project so it will take me way too long to convert that over to something else. So yeah if development was easy and all the plugins worked well together in dreamland, the transfer over to a headless framework with all the bells and whistles that we needed with the least amount of prep for the API (a.k.a. CraftQL) would have been great. But it looks like my luck has run out this time, unfortunately. 

Thank you very much for your help in this regard. I will be using this plugin in future projects where I do not need to use NEO again. 

Hope you have a great day further. :)

**Comment by narration-sd** [comment]
Thank, and you too :)

Here's a thought. How about just using EApi/Neo where you can't root it out (one page or so, we could hope??), and formatting the rest of the upgrade using CraftQL?

Would smooth a way forward, maybe...anyway, best fortune, and cheers...

**Comment by dineufeld** [comment]
If somebody looking for alternative for your next project take look at WordPress Advanced Custom Fields and new released GraphQL Plugin: [WPGraphQL](https://github.com/wp-graphql/wp-graphql-acf). If your project requires flexibility and modular layout it my be worth it.
